File: app/src/main/java/com/pep1lo/bakerspercentagecalculator/MainActivity.java
package com.pep1lo.bakerspercentagecalculator;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.PopupMenu;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.pep1lo.bakerspercentagecalculator.R;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.gson.ExclusionStrategy;
import com.google.gson.FieldAttributes;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.google.zxing.BinaryBitmap;
import com.google.zxing.LuminanceSource;
import com.google.zxing.MultiFormatReader;
import com.google.zxing.RGBLuminanceSource;
import com.google.zxing.Reader;
import com.google.zxing.Result;
import com.google.zxing.common.HybridBinarizer;
import com.journeyapps.barcodescanner.ScanContract;
import com.journeyapps.barcodescanner.ScanOptions;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private RecipeDataSource dataSource;
    private RecyclerView recyclerViewRecipes;
    private FloatingActionButton fabAddRecipe;
    private RecipeAdapter adapter;
    private List<Recipe> recipes = new ArrayList<>();

    private final ActivityResultLauncher<Intent> addRecipeLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == RESULT_OK) {
                    loadRecipes();
                }
            });

    private final ActivityResultLauncher<Intent> backupLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
                    Uri uri = result.getData().getData();
                    writeBackupToFile(uri);
                }
            });

    private final ActivityResultLauncher<Intent> restoreLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
                    Uri uri = result.getData().getData();
                    readBackupFromFile(uri);
                }
            });

    // Launcher for the QR Code Camera Scanner
    private final ActivityResultLauncher<ScanOptions> qrCodeScannerLauncher = registerForActivityResult(new ScanContract(),
            result -> {
                if(result.getContents() != null) {
                    importRecipeFromJson(result.getContents());
                }
            });

    // New launcher for picking an image from the gallery
    private final ActivityResultLauncher<Intent> pickImageLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
                    Uri imageUri = result.getData().getData();
                    scanQrCodeFromImage(imageUri);
                }
            });


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        com.google.android.material.appbar.MaterialToolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        dataSource = RecipeDataSource.getInstance(this);
        dataSource.open();

        recyclerViewRecipes = findViewById(R.id.recyclerViewRecipes);
        fabAddRecipe = findViewById(R.id.fabAddRecipe);

        setupRecyclerView();
        loadRecipes();

        fabAddRecipe.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, AddEditRecipeActivity.class);
            addRecipeLauncher.launch(intent);
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_backup) {
            initiateBackup();
            return true;
        } else if (id == R.id.action_restore) {
            initiateRestore();
            return true;
        } else if (id == R.id.action_import_recipe) {
            showImportDialog();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void showImportDialog() {
        final CharSequence[] options = {
                getString(R.string.scan_with_camera),
                getString(R.string.choose_from_gallery)
        };
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.import_recipe);
        builder.setItems(options, (dialog, which) -> {
            if (options[which].equals(getString(R.string.scan_with_camera))) {
                initiateScan();
            } else if (options[which].equals(getString(R.string.choose_from_gallery))) {
                initiateImagePick();
            }
        });
        builder.show();
    }

    private void initiateImagePick() {
        Intent intent = new Intent(Intent.ACTION_PICK);
        intent.setType("image/*");
        pickImageLauncher.launch(intent);
    }

    private void scanQrCodeFromImage(Uri imageUri) {
        try {
            InputStream inputStream = getContentResolver().openInputStream(imageUri);
            Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
            if (bitmap == null) {
                Toast.makeText(this, R.string.qr_code_not_found, Toast.LENGTH_SHORT).show();
                return;
            }
            int[] intArray = new int[bitmap.getWidth() * bitmap.getHeight()];
            bitmap.getPixels(intArray, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(), bitmap.getHeight());
            LuminanceSource source = new RGBLuminanceSource(bitmap.getWidth(), bitmap.getHeight(), intArray);
            BinaryBitmap binaryBitmap = new BinaryBitmap(new HybridBinarizer(source));
            Reader reader = new MultiFormatReader();
            Result result = reader.decode(binaryBitmap);
            importRecipeFromJson(result.getText());
        } catch (Exception e) {
            Toast.makeText(this, R.string.qr_code_not_found, Toast.LENGTH_SHORT).show();
            e.printStackTrace();
        }
    }

    private void initiateScan() {
        ScanOptions options = new ScanOptions();
        options.setDesiredBarcodeFormats(ScanOptions.QR_CODE);
        options.setPrompt("Scan a recipe QR code");
        options.setCameraId(0);  // Use a specific camera of the device
        options.setBeepEnabled(true);
        options.setBarcodeImageEnabled(true);
        qrCodeScannerLauncher.launch(options);
    }

    private void initiateBackup() {
        Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("application/json");
        intent.putExtra(Intent.EXTRA_TITLE, "bakers_percentage_backup.json");
        backupLauncher.launch(intent);
    }

    private void writeBackupToFile(Uri uri) {
        try {
            Gson gson = new GsonBuilder().setPrettyPrinting().create();
            String json = gson.toJson(recipes);
            getContentResolver().openOutputStream(uri).write(json.getBytes());
            Toast.makeText(this, R.string.backup_successful, Toast.LENGTH_SHORT).show();
        } catch (Exception e) {
            Toast.makeText(this, R.string.backup_failed, Toast.LENGTH_SHORT).show();
        }
    }

    private void initiateRestore() {
        new MaterialAlertDialogBuilder(this)
                .setTitle(R.string.restore_confirmation_title)
                .setMessage(R.string.restore_confirmation_message)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.restore, (dialog, which) -> {
                    Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
                    intent.addCategory(Intent.CATEGORY_OPENABLE);
                    intent.setType("application/json");
                    restoreLauncher.launch(intent);
                })
                .show();
    }

    private void importRecipeFromJson(String jsonString) {
        Gson gson = new Gson();
        try {
            Recipe recipe = gson.fromJson(jsonString, Recipe.class);

            if (recipe != null && recipe.getName() != null && !recipe.getName().isEmpty()) {
                dataSource.createRecipe(recipe);
                loadRecipes();
                Toast.makeText(this, R.string.recipe_imported_successfully, Toast.LENGTH_SHORT).show();
            } else {
                throw new Exception("Invalid recipe format.");
            }
        } catch (Exception e) {
            new MaterialAlertDialogBuilder(this)
                    .setTitle(R.string.import_failed)
                    .setMessage(R.string.invalid_recipe_format)
                    .setPositiveButton(android.R.string.ok, null)
                    .show();
        }
    }

    private void readBackupFromFile(Uri uri) {
        try {
            InputStream inputStream = getContentResolver().openInputStream(uri);
            BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
            StringBuilder stringBuilder = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                stringBuilder.append(line);
            }
            inputStream.close();

            Gson gson = new Gson();
            Type listType = new TypeToken<ArrayList<Recipe>>(){}.getType();
            List<Recipe> restoredRecipes = gson.fromJson(stringBuilder.toString(), listType);

            if (restoredRecipes != null) {
                dataSource.deleteAllRecipes();
                for (Recipe recipe : restoredRecipes) {
                    dataSource.createRecipe(recipe);
                }
            }

            loadRecipes();
            Toast.makeText(this, R.string.restore_successful, Toast.LENGTH_SHORT).show();
        } catch (Exception e) {
            Toast.makeText(this, R.string.restore_failed, Toast.LENGTH_SHORT).show();
        }
    }

    private void setupRecyclerView() {
        adapter = new RecipeAdapter(recipes,
                recipe -> {
                    Intent intent = new Intent(MainActivity.this, CalculateActivity.class);
                    intent.putExtra("recipe", recipe);
                    startActivity(intent);
                },
                this::showDeleteConfirmationDialog,
                this::showShareMenu
        );
        recyclerViewRecipes.setLayoutManager(new LinearLayoutManager(this));
        recyclerViewRecipes.setAdapter(adapter);
    }

    private void showShareMenu(Recipe recipe, View anchor) {
        PopupMenu popup = new PopupMenu(this, anchor);
        popup.getMenuInflater().inflate(R.menu.share_options_menu, popup.getMenu());
        popup.setOnMenuItemClickListener(item -> {
            int itemId = item.getItemId();
            if (itemId == R.id.action_share_text) {
                shareRecipeAsText(recipe);
                return true;
            } else if (itemId == R.id.action_share_qr) {
                shareRecipeAsQr(recipe);
                return true;
            }
            return false;
        });
        popup.show();
    }

    private void shareRecipeAsText(Recipe recipe) {
        StringBuilder shareText = new StringBuilder();
        shareText.append(getString(R.string.recipe_share_header, recipe.getName())).append("\n\n");
        shareText.append(getString(R.string.bakers_percentages_header)).append("\n");

        for (Ingredient ingredient : recipe.getIngredients()) {
            shareText.append("- ").append(ingredient.getName()).append(": ").append(ingredient.getWeight()).append("%\n");
        }

        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.bakers_recipe_subject, recipe.getName()));
        shareIntent.putExtra(Intent.EXTRA_TEXT, shareText.toString());

        startActivity(Intent.createChooser(shareIntent, getString(R.string.share_recipe_via)));
    }

    private void shareRecipeAsQr(Recipe recipe) {
        // This strategy tells Gson to skip the 'id' and 'lastTotalWeight' fields during serialization
        ExclusionStrategy strategy = new ExclusionStrategy() {
            @Override
            public boolean shouldSkipField(FieldAttributes f) {
                return f.getName().equals("id") || f.getName().equals("lastTotalWeight");
            }

            @Override
            public boolean shouldSkipClass(Class<?> clazz) {
                return false;
            }
        };

        Gson gson = new GsonBuilder()
                .addSerializationExclusionStrategy(strategy)
                .create();

        String json = gson.toJson(recipe);

        Intent intent = new Intent(this, DisplayQrActivity.class);
        intent.putExtra("RECIPE_JSON", json);
        startActivity(intent);
    }

    private void loadRecipes() {
        recipes = dataSource.getAllRecipes();
        adapter.setRecipes(recipes);
    }

    private void showDeleteConfirmationDialog(Recipe recipe) {
        new MaterialAlertDialogBuilder(this)
                .setTitle(R.string.delete_confirmation_title)
                .setMessage(getString(R.string.delete_confirmation_message, recipe.getName()))
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.delete, (dialog, which) -> {
                    deleteRecipe(recipe);
                })
                .show();
    }

    private void deleteRecipe(Recipe recipe) {
        dataSource.deleteRecipe(recipe.getId());
        loadRecipes();
    }

    @Override
    protected void onResume() {
        super.onResume();
        loadRecipes();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        dataSource.close();
    }
}

File: app/src/main/java/com/pep1lo/bakerspercentagecalculator/MainActivity.kt
package com.pep1lo.bakerspercentagecalculator

import android.app.Activity
import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.PopupMenu
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.gson.ExclusionStrategy
import com.google.gson.FieldAttributes
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.google.zxing.BinaryBitmap
import com.google.zxing.MultiFormatReader
import com.google.zxing.RGBLuminanceSource
import com.google.zxing.common.HybridBinarizer
import com.journeyapps.barcodescanner.ScanContract
import com.journeyapps.barcodescanner.ScanOptions
import com.pep1lo.bakerspercentagecalculator.databinding.ActivityMainBinding
import kotlinx.coroutines.launch
import java.io.BufferedReader
import java.io.InputStreamReader

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var dataSource: RecipeDataSource
    private lateinit var adapter: RecipeAdapter
    private var recipes: List<Recipe> = ArrayList()

    private val addRecipeLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == RESULT_OK) {
            loadRecipes()
        }
    }

    private val backupLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK && result.data != null) {
            result.data?.data?.let { uri ->
                writeBackupToFile(uri)
            }
        }
    }

    private val restoreLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK && result.data != null) {
            result.data?.data?.let { uri ->
                readBackupFromFile(uri)
            }
        }
    }

    private val qrCodeScannerLauncher = registerForActivityResult(ScanContract()) { result ->
        if (result.contents != null) {
            importRecipeFromJson(result.contents)
        }
    }

    private val pickImageLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK && result.data != null) {
            result.data?.data?.let { imageUri ->
                scanQrCodeFromImage(imageUri)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)

        dataSource = RecipeDataSource.getInstance(this)
        try {
            dataSource.open()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        setupRecyclerView()
        loadRecipes()

        binding.fabAddRecipe.setOnClickListener {
            val intent = Intent(this@MainActivity, AddEditRecipeActivity::class.java)
            addRecipeLauncher.launch(intent)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_backup -> {
                initiateBackup()
                true
            }
            R.id.action_restore -> {
                initiateRestore()
                true
            }
            R.id.action_import_recipe -> {
                showImportDialog()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun showImportDialog() {
        val options = arrayOf<CharSequence>(
            getString(R.string.scan_with_camera),
            getString(R.string.choose_from_gallery)
        )
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.import_recipe)
        builder.setItems(options) { _, which ->
            if (options[which] == getString(R.string.scan_with_camera)) {
                initiateScan()
            } else if (options[which] == getString(R.string.choose_from_gallery)) {
                initiateImagePick()
            }
        }
        builder.show()
    }

    private fun initiateImagePick() {
        val intent = Intent(Intent.ACTION_PICK)
        intent.type = "image/*"
        pickImageLauncher.launch(intent)
    }

    private fun scanQrCodeFromImage(imageUri: Uri) {
        try {
            val inputStream = contentResolver.openInputStream(imageUri)
            val bitmap = BitmapFactory.decodeStream(inputStream)
            if (bitmap == null) {
                Toast.makeText(this, R.string.qr_code_not_found, Toast.LENGTH_SHORT).show()
                return
            }
            val intArray = IntArray(bitmap.width * bitmap.height)
            bitmap.getPixels(intArray, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
            val source = RGBLuminanceSource(bitmap.width, bitmap.height, intArray)
            val binaryBitmap = BinaryBitmap(HybridBinarizer(source))
            val reader = MultiFormatReader()
            val result = reader.decode(binaryBitmap)
            importRecipeFromJson(result.text)
        } catch (e: Exception) {
            Toast.makeText(this, R.string.qr_code_not_found, Toast.LENGTH_SHORT).show()
            e.printStackTrace()
        }
    }

    private fun initiateScan() {
        val options = ScanOptions()
        options.setDesiredBarcodeFormats(ScanOptions.QR_CODE)
        options.setPrompt("Scan a recipe QR code")
        options.setCameraId(0)
        options.setBeepEnabled(true)
        options.setBarcodeImageEnabled(true)
        qrCodeScannerLauncher.launch(options)
    }

    private fun initiateBackup() {
        val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
        intent.addCategory(Intent.CATEGORY_OPENABLE)
        intent.type = "application/json"
        intent.putExtra(Intent.EXTRA_TITLE, "bakers_percentage_backup.json")
        backupLauncher.launch(intent)
    }

    private fun writeBackupToFile(uri: Uri) {
        lifecycleScope.launch {
            try {
                val gson = GsonBuilder().setPrettyPrinting().create()
                val json = gson.toJson(recipes)
                contentResolver.openOutputStream(uri)?.write(json.toByteArray())
                Toast.makeText(this@MainActivity, R.string.backup_successful, Toast.LENGTH_SHORT).show()
            } catch (e: Exception) {
                Toast.makeText(this@MainActivity, R.string.backup_failed, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun initiateRestore() {
        MaterialAlertDialogBuilder(this)
            .setTitle(R.string.restore_confirmation_title)
            .setMessage(R.string.restore_confirmation_message)
            .setNegativeButton(R.string.cancel, null)
            .setPositiveButton(R.string.restore) { _, _ ->
                val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                intent.type = "application/json"
                restoreLauncher.launch(intent)
            }
            .show()
    }

    private fun importRecipeFromJson(jsonString: String) {
        val gson = Gson()
        lifecycleScope.launch {
            try {
                val recipe = gson.fromJson(jsonString, Recipe::class.java)

                if (recipe != null && !recipe.name.isNullOrEmpty()) {
                    dataSource.createRecipe(recipe)
                    loadRecipes()
                    Toast.makeText(this@MainActivity, R.string.recipe_imported_successfully, Toast.LENGTH_SHORT).show()
                } else {
                    throw Exception("Invalid recipe format.")
                }
            } catch (e: Exception) {
                MaterialAlertDialogBuilder(this@MainActivity)
                    .setTitle(R.string.import_failed)
                    .setMessage(R.string.invalid_recipe_format)
                    .setPositiveButton(android.R.string.ok, null)
                    .show()
            }
        }
    }

    private fun readBackupFromFile(uri: Uri) {
        lifecycleScope.launch {
            try {
                val inputStream = contentResolver.openInputStream(uri)
                val reader = BufferedReader(InputStreamReader(inputStream))
                val stringBuilder = StringBuilder()
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    stringBuilder.append(line)
                }
                inputStream?.close()

                val gson = Gson()
                val listType = object : TypeToken<ArrayList<Recipe>>() {}.type
                val restoredRecipes: List<Recipe> = gson.fromJson(stringBuilder.toString(), listType)

                if (restoredRecipes != null) {
                    dataSource.deleteAllRecipes()
                    for (recipe in restoredRecipes) {
                        dataSource.createRecipe(recipe)
                    }
                }

                loadRecipes()
                Toast.makeText(this@MainActivity, R.string.restore_successful, Toast.LENGTH_SHORT).show()
            } catch (e: Exception) {
                Toast.makeText(this@MainActivity, R.string.restore_failed, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun setupRecyclerView() {
        adapter = RecipeAdapter(recipes,
            { recipe ->
                val intent = Intent(this@MainActivity, CalculateActivity::class.java)
                intent.putExtra("recipe", recipe)
                startActivity(intent)
            },
            { recipe -> showDeleteConfirmationDialog(recipe) },
            { recipe, view -> showShareMenu(recipe, view) }
        )
        binding.recyclerViewRecipes.layoutManager = LinearLayoutManager(this)
        binding.recyclerViewRecipes.adapter = adapter
    }

    private fun showShareMenu(recipe: Recipe, anchor: View) {
        val popup = PopupMenu(this, anchor)
        popup.menuInflater.inflate(R.menu.share_options_menu, popup.menu)
        popup.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_share_text -> {
                    shareRecipeAsText(recipe)
                    true
                }
                R.id.action_share_qr -> {
                    shareRecipeAsQr(recipe)
                    true
                }
                else -> false
            }
        }
        popup.show()
    }

    private fun shareRecipeAsText(recipe: Recipe) {
        val shareText = StringBuilder()
        shareText.append(getString(R.string.recipe_share_header, recipe.name)).append("\n\n")
        shareText.append(getString(R.string.bakers_percentages_header)).append("\n")

        for (ingredient in recipe.ingredients) {
            shareText.append("- ").append(ingredient.name).append(": ").append(ingredient.weight).append("%\n")
        }

        val shareIntent = Intent(Intent.ACTION_SEND)
        shareIntent.type = "text/plain"
        shareIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.bakers_recipe_subject, recipe.name))
        shareIntent.putExtra(Intent.EXTRA_TEXT, shareText.toString())

        startActivity(Intent.createChooser(shareIntent, getString(R.string.share_recipe_via)))
    }

    private fun shareRecipeAsQr(recipe: Recipe) {
        val strategy = object : ExclusionStrategy {
            override fun shouldSkipField(f: FieldAttributes): Boolean {
                return f.name == "id" || f.name == "lastTotalWeight"
            }

            override fun shouldSkipClass(clazz: Class<*>?): Boolean {
                return false
            }
        }

        val gson = GsonBuilder()
            .addSerializationExclusionStrategy(strategy)
            .create()

        val json = gson.toJson(recipe)

        val intent = Intent(this, DisplayQrActivity::class.java)
        intent.putExtra("RECIPE_JSON", json)
        startActivity(intent)
    }

    private fun loadRecipes() {
        lifecycleScope.launch {
            recipes = dataSource.getAllRecipes()
            adapter.setRecipes(recipes)

            if (recipes.isEmpty()) {
                binding.recyclerViewRecipes.visibility = View.GONE
                binding.textViewEmptyState.visibility = View.VISIBLE
            } else {
                binding.recyclerViewRecipes.visibility = View.VISIBLE
                binding.textViewEmptyState.visibility = View.GONE
            }
        }
    }

    private fun showDeleteConfirmationDialog(recipe: Recipe) {
        MaterialAlertDialogBuilder(this)
            .setTitle(R.string.delete_confirmation_title)
            .setMessage(getString(R.string.delete_confirmation_message, recipe.name))
            .setNegativeButton(R.string.cancel, null)
            .setPositiveButton(R.string.delete) { _, _ ->
                deleteRecipe(recipe)
            }
            .show()
    }

    private fun deleteRecipe(recipe: Recipe) {
        lifecycleScope.launch {
            dataSource.deleteRecipe(recipe.id)
            loadRecipes()
        }
    }

    override fun onResume() {
        super.onResume()
        loadRecipes()
    }

    override fun onDestroy() {
        super.onDestroy()
        dataSource.close()
    }
}
